gunicorn_thrift/thriftpy_gevent_worker.py

```diff
@@ -52,6 +75,79 @@ def check_protocol_and_transport(app):
+            self._active_hub = gevent.hub.get_hub()
+            # Set up a trace function to record each greenlet switch.
+            self._active_greenlet = None
+            self._greenlet_switch_counter = 0
+            greenlet.settrace(self._greenlet_switch_tracer)
+            self._main_thread_id = _real_get_ident()
+            # Create a real thread to monitor out execution.
+            # Since this will be a long-running daemon thread, it's OK to
+            # fire-and-forget using the low-level start_new_thread function.
+            _real_start_new_thread(self._process_monitoring_thread, ())
+
+        return super(GeventThriftPyWorker, self).init_process()
+
+    def _greenlet_switch_tracer(self, what, (origin, target)):
+        """Callback method executed on every greenlet switch.
+
+        The worker arranges for this method to be called on every greenlet
+        switch.  It keeps track of which greenlet is currently active and
+        increments a counter to track how many switches have been performed.
+        """
+        # Increment the counter to indicate that a switch took place.
+        # This will periodically be reset to zero by the monitoring thread,
+        # so we don't need to worry about it growing without bound.
+        self._active_greenlet = target
+        self._greenlet_switch_counter += 1
+
+    def _process_monitoring_thread(self):
+        """Method run in background thread that monitors our execution.
+
+        This method is an endless loop that gets executed in a background
+        thread.  It periodically wakes up and checks:
+
+            * whether the active greenlet has switched since last checked
+
+        """
+        try:
+            while True:
+                _real_sleep(self.app.cfg.gevent_check_interval)
+                self._check_greenlet_blocking()
+        except Exception:
+            # Swallow any exceptions raised during interpreter shutdown.
+            # Daemonic Thread objects have this same behaviour.
+            if sys is not None:
+                raise
+
+    def _check_greenlet_blocking(self):
+        if not self.app.cfg.gevent_check_interval:
+            return
+        # If there have been no greenlet switches since we last checked,
+        # grab the stack trace and log an error.  The active greenlet's frame
+        # is not available from the greenlet object itself, we have to look
+        # up the current frame of the main thread for the traceback.
+        if self._greenlet_switch_counter == 0:
+            active_greenlet = self._active_greenlet
+            # The hub gets a free pass, since it blocks waiting for IO.
+            if active_greenlet not in (None, self._active_hub):
+                frame = sys._current_frames()[self._main_thread_id]
+                stack = traceback.format_stack(frame)
+                err_log = ["Greenlet appears to be blocked\n"] + stack
```

Comment:
Shall we `delete frame` here to avoid from potential memory leaking?

Reply:
But I don't see how this could be leaked...

Reply:
~~Is it possible that `frame` is pointed to current frame and builds a circular reference?~~

Yes, it could not be leaked. The `self._main_thread_id` will never match the current frame's thread.